src/Build/Definition/Project.cs

```diff
@@ -1117,32 +1129,161 @@ public List<GlobResult> GetAllGlobs(string itemType)
+                excludeGlob = excludeItemspec.ToMSBuildGlob();
+            }
+
+            IEnumerable<string> removeFragmentStrings = Enumerable.Empty<string>();
+            IMSBuildGlob removeGlob = null;
+
+            if (removeElementCache.ContainsKey(itemElement.ItemType))
+            {
+                removeFragmentStrings = removeElementCache[itemElement.ItemType].FragmentStrings.ToImmutable();
+                removeGlob = new CompositeGlob(removeElementCache[itemElement.ItemType].Globs.ToImmutable());
+            }
+
+            var includeGlobWithGaps = CreateIncludeGlobWithGaps(includeGlob, excludeGlob, removeGlob);
+
+            return new GlobResult(itemElement, includeGlobStrings, includeGlobWithGaps, excludeFragmentStrings, removeFragmentStrings);
+        }
+
+        private static IMSBuildGlob CreateIncludeGlobWithGaps(IMSBuildGlob includeGlob, IMSBuildGlob excludeGlob, IMSBuildGlob removeGlob)
+        {
+            if (excludeGlob != null && removeGlob != null)
+            {
+                return new MSBuildGlobWithGaps(
+                    includeGlob,
+                    new CompositeGlob(
+                        excludeGlob,
+                        removeGlob
+                    ));
+            }
+
+            if (excludeGlob != null || removeGlob != null)
+            {
+                var gapGlob = excludeGlob ?? removeGlob;
+
+                return new MSBuildGlobWithGaps(
+                    includeGlob,
+                    gapGlob
+                );
+            }
+
+            return includeGlob;
+        }
+
+        private void CacheInformationFromRemoveItem(ProjectItemElement itemElement, Dictionary<string, CumulatedRemoveElementData> removeElementCache)
+        {
+            CumulatedRemoveElementData cumulatedRemoveElementData;
+            if (!removeElementCache.TryGetValue(itemElement.ItemType, out cumulatedRemoveElementData))
+            {
+                cumulatedRemoveElementData = new CumulatedRemoveElementData
+                {
+                    Globs = ImmutableList.CreateBuilder<IMSBuildGlob>(),
+                    FragmentStrings = ImmutableHashSet.CreateBuilder<string>()
+                };
+
+                removeElementCache[itemElement.ItemType] = cumulatedRemoveElementData;
+            }
+
+            var removeSpec = new EvaluationItemSpec(itemElement.Remove, _data.Expander, itemElement.RemoveLocation);
+            var removeSpecFragmentStrings = GetRelevantFragmentStringsForExcludesAndRemoves(removeSpec);
+            var removeGlob = removeSpec.ToMSBuildGlob();
+
+            cumulatedRemoveElementData.Globs.Add(removeGlob);
+
+            foreach (var removeFragment in removeSpecFragmentStrings)
+            {
+                cumulatedRemoveElementData.FragmentStrings.Add(removeFragment);
+            }
+        }
+
+        private IEnumerable<string> GetRelevantFragmentStringsForExcludesAndRemoves(EvaluationItemSpec spec)
```

Comment:
Needs documentation:
* Reorders the fragments
* What fragments does it not return?
* "Relevant" to what?